krm-functions/ipam-fn/fn/function.go

```diff
@@ -0,0 +1,84 @@
+/*
+ Copyright 2023 The Nephio Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package fn
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	"github.com/nephio-project/nephio/krm-functions/lib/condkptsdk"
+	"github.com/nephio-project/nephio/krm-functions/lib/kubeobject"
+	ipamv1alpha1 "github.com/nokia/k8s-ipam/apis/alloc/ipam/v1alpha1"
+	"github.com/nokia/k8s-ipam/pkg/proxy/clientproxy"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type FnR struct {
+	ClientProxy clientproxy.Proxy[*ipamv1alpha1.NetworkInstance, *ipamv1alpha1.IPAllocation]
+}
+
+func (r *FnR) Run(rl *fn.ResourceList) (bool, error) {
+	sdk, err := condkptsdk.New(
+		rl,
+		&condkptsdk.Config{
+			For: corev1.ObjectReference{
+				APIVersion: ipamv1alpha1.GroupVersion.Identifier(),
+				Kind:       ipamv1alpha1.IPAllocationKind,
+			},
+			PopulateOwnResourcesFn: nil,
+			GenerateResourceFn:     r.updateIPAllocationResource,
+		},
+	)
+	if err != nil {
+		rl.Results.ErrorE(err)
+		return false, nil
+	}
+	return sdk.Run()
+}
+
+// updateIPAllocationResource provides an ip allocation for a given IPAllocation KRM resource
+// in the package by calling the ipam backend
+func (r *FnR) updateIPAllocationResource(forObj *fn.KubeObject, objs fn.KubeObjects) (*fn.KubeObject, error) {
+	if forObj == nil {
+		return nil, fmt.Errorf("expected a for object but got nil")
+	}
+	fn.Logf("ipalloc: %v\n", forObj)
+	allocKOE, err := kubeobject.NewFromKubeObject[ipamv1alpha1.IPAllocation](forObj)
+	if err != nil {
+		return nil, err
+	}
+	alloc, err := allocKOE.GetGoStruct()
+	if err != nil {
+		return nil, err
+	}
+	resp, err := r.ClientProxy.Allocate(context.Background(), alloc, nil)
```

Comment:
If I understand correctly this is intended to use both with a mock backend proxy, but also with an actual backend, as well. The later case seems to contradict with the "lambda function" nature of KRM functions: it may generate different results for the same input.
I understand why this function is useful for testing purposes, but for calling out to the real backend, shouldn't we  use a Kubernetes Operator, more precisely a Specializer instead?

Reply:
My understanding is that the Specializer should watch the kpt package and detect xxxAllocations with an empty `status` automatically, without the help of a KRM function. Isn't it the case?

Reply:
this is the call to the Grpc call to the backend. As long as the backend does not get updated the result will be the same. So I don't see the contradiction.

Reply:
I exposed the fn because we can run them independently for our porch independent test pipeline. But in a real environment this will run as a specialised controller and not through the kpt pipeline.

Reply:
yes, that was my question, thanks for clarifying.